File: Week6/homework/sentimental-credit/credit.py
from cs50 import get_int
from math import floor

card_number = get_int("What is your credit card number?")

# Amex 34, 37 and 15 digits
# Mastercard 51-55 and 16 digits
# Visa 4 and 13 or 16 digits

sum = 0

number = card_number

while number >= 1:
    digit1 = int(number % 10)
    digit2 = int((number / 10) % 10)
    product_digits = int(((digit2 * 2) % 10) + floor(digit2 * 2 / 10))
    sum = sum + product_digits + digit1
    number = int(number / 100)

print(f"{sum}")

digits = str(card_number)
length = len(digits)
print(f"digits are {digits[0]}, length is {length}")
mastercard = ["1", "2", "3", "4", "5"]

if sum % 10 == 0 and digits[0] == "4" and (length == 16 or length == 13):
    print("VISA\n")
elif sum % 10 == 0 and digits[0] == "3" and (digits[1] == "4" or digits[1] == "7") and length == 15:
    print("AMEX\n")
elif sum % 10 == 0 and digits[0] == "5" and digits[1] in mastercard and length == 16:
    print("MASTERCARD\n")
else:
    print("INVALID\n")

File: Week6/homework/sentimental-credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Amex 34, 37
// Mastercard 51, 52, 53, 54, 55
// Visa 4

int main(void)
{
    // get the card number
    long card_number = get_long("What is your credit card number?");

    // find the number of digits
    int count = floor(log10(card_number) + 1);

    // implement Luhn's formula
    int sum = 0;

    long number = card_number;

    while (number != 0)
    {
        int digit1 = number % 10;        // produces the remainder of dividing by 10
        int digit2 = (number / 10) % 10; // producing the remainder of the updated number
        digit2 = digit2 * 2;
        int product_digits = (digit2 % 10) + floor(digit2 / 10);
        sum = sum + digit1 + product_digits;
        number = number / 100;
        // printf("sum is %i, digit1 is %i, digit 2 is %i\n", sum, digit1, digit2);
    }

    // printf("end sum is %i\n", sum);

    // report the results
    char digits[count];
    sprintf(digits, "%li", card_number);
    int length = strlen(digits);
    printf("%i\n", length);

    int visa = 4;
    int amex = 3;
    int master = 5;
    int first_digit = digits[0] - '0';
    int second_digit = digits[1] - '0';

    if (sum % 10 == 0 && visa == first_digit && (length == 16 || length == 13))
    {
        printf("VISA\n");
    }
    else if (sum % 10 == 0 && amex == first_digit && second_digit == (4 | 7) && length == 15)
    {
        printf("AMEX\n");
    }
    else if (sum % 10 == 0 && master == first_digit && second_digit >= 1 && second_digit <= 5 &&
             length == 16)
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
